File: kubicekkocnar.client/src/types/Block.ts
import { Material } from 'three';

// a Block is a type referenced in the PlacedBlock type
// used to define the properties of a rendered block in the game
interface Block {
    blockId: number;
    name: string;

    attributes: string[]; //works like classes in HTML -> needs conversion from [server] string to [client] string[]
    
    material?: Material | Material[]; // asigned once the first block of this type is added to the scene

    texture0Id: number; // when only texture0 is set, it is used for all sides 
    texture1Id: number;
    texture2Id: number;
    texture3Id: number;
    texture4Id: number;
    texture5Id: number; // when texture5 is set, all textures are used for each side

    created: Date;
}

export default Block;

File: KubicekKocnar.Server/Models/Block.cs
using Newtonsoft.Json;

namespace KubicekKocnar.Server.Models
{
    public class Block
    {
        public uint BlockId { get; set; }

        public required string Name { get; set; }
        public required DateTime Created { get; set; } = DateTime.Now;

        public uint? Texture0Id { get; set; }
        [JsonIgnore]
        public Texture? Texture0 { get; set; }

        public uint? Texture1Id { get; set; }
        [JsonIgnore]
        public Texture? Texture1 { get; set; }

        public uint? Texture2Id { get; set; }
        [JsonIgnore]
        public Texture? Texture2 { get; set; }

        public uint? Texture3Id { get; set; }
        [JsonIgnore]
        public Texture? Texture3 { get; set; 
        }
        public uint? Texture4Id { get; set; }
        [JsonIgnore]
        public Texture? Texture4 { get; set; }

        public uint? Texture5Id { get; set; }
        [JsonIgnore]
        public Texture? Texture5 { get; set; }

        public required string Attributes { get; set; } = "";

        [JsonIgnore]
        public ICollection<PlacedBlock> PlacedBlocks { get; set; } = [];

    }
}
